### ISSUE TITLE
ContainerPort protocol is a key field, which makes it required for server-side apply

### ISSUE THREAD

**Issue by bouk** [kubernetes-sigs/structured-merge-diff#130]
Hi,

I was playing around with server-side apply and hit into a problem. Say I have a pod like this (very reduced case of course):

```yaml
# pod.yml
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  containers:
  - name: test 
    ports:
    - containerPort: 1234
```

If I run `kubectl apply -f pod.yml --server-side` I get the following error:

`Error from server: failed to create typed patch object: .spec.containers[name="test"].ports: element 0: associative list with keys has an element that omits key field "protocol"`

This is unexpected, because [ContainerPort](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#containerport-v1-core) specifies the protocol as an optional field, which defaults to `TCP`. I traced this error back to this repository, so I figured this is where it could be resolved.

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T19:18:23Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-11-16T01:01:59Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by bouk** [comment]
I think keeping this restriction in will hamper the adoption of server-side applies greatly, for example the ingress-nginx config also misses the 'protocol' ports in its config: https://github.com/kubernetes/ingress-nginx/blob/master/deploy/static/provider/cloud-generic.yaml

cc @lavalamp

**Comment by apelisse** [comment]
Thanks @bouk, we're aware of this bug and we do want to fix it, thanks.

We have #83 that partially fixes it (still trying to figure out the best solution).

/assign @jennybuckley

**Comment by apelisse** [comment]
/reopen

We care about this more than ever.

**Comment by anwarchk** [comment]
Any idea when this would be fixed ? We are also adapting Server Side apply, but we have a lot of helm charts that doesn't have protocol specified in the port section.

**Comment by apelisse** [comment]
This is planned for 1.20

**Comment by apelisse** [comment]
We're actively working on this!

**Comment by rocketraman** [comment]
Despite https://github.com/kubernetes/ingress-nginx/pull/4791, the nginx controller still does not specify the protocol in all its configurations, so this still prevents `--server-side` from being used. See for example: https://github.com/kubernetes/ingress-nginx/blob/master/charts/ingress-nginx/templates/controller-service-webhook.yaml.

**Comment by apelisse** [comment]
Should be fixed in 1.20!

**Comment by codablock** [comment]
As I've deployed my first 1.20.2 cluster today, I retried server-side applying a Service without the protocol field being set and I still have the same issue. @apelisse Do you have a link to the PR that fixed this in 1.20?

**Comment by liggitt** [comment]
this was narrowly fixed for container port in 1.20. the service port default is addressed in master in https://github.com/kubernetes/kubernetes/pull/98576/commits/dab511224020f6f7a896560d7c48770187a683c1

**Comment by apelisse** [comment]
> the service port default is addressed in master in kubernetes/kubernetes@dab5112

@liggitt What do you think about backporting this to 1.20?
